### PR TITLE
Allow registering header/footer sections in the file lists

### PR DESCRIPTION
**Pull request by juliusknorr**
Allows filelist plugins to register header/footer sections

- Text https://github.com/nextcloud/text/pull/351
- Recommendations https://github.com/nextcloud/recommendations/pull/136

### CODE REVIEW

File: apps/files/js/filelist.js [hunk 1]
```diff
@@ -408,6 +416,40 @@
...
+
+		initHeadersAndFooters: function() {
+			this.headers.sort(function(a, b) {
+				return a.order - b.order;
+			})
+			this.footers.sort(function(a, b) {
+				return a.order - b.order;
+			})
+			var uniqueIds = [];
+			this.headers.forEach((header) => {
```
Comment (kesselb) [inline apps/files/js/filelist.js]: Is `filelist.js` tanspiled by babel?

Comment (juliushaertl) [reply apps/files/js/filelist.js]: :see_no_evil: Ah of course not, thanks

Comment (kesselb) [reply apps/files/js/filelist.js]: Just to be sure. There is a forEach polyfill? Not that I care about IE11 but ...

Comment (juliushaertl) [reply apps/files/js/filelist.js]: forEach is supported since IE9 so no polyfill needed for that.

Comment (kesselb) [reply apps/files/js/filelist.js]: :tada: I checked the wrong forEach obviously :see_no_evil:

File: apps/files/js/filelist.js [hunk 2]
```diff
@@ -408,6 +416,46 @@
...
+			this.footers.forEach(function(footer) {
+				if (footer.id) {
+					if (uniqueIds.indexOf(footer.id) !== -1) {
+						return
+					}
+					uniqueIds.push(footer.id)
+				}
+				self.$footer.append(footer.el)
+				setTimeout(function() {
+					header.footer(self)
```
Comment (danxuliu) [inline apps/files/js/filelist.js]: I guess this should be `footer.render(self)` instead.

File: apps/files/js/filelist.js [hunk 3]
```diff
@@ -3642,6 +3690,18 @@
...
+
+		registerHeader: function(header) {
+			this.headers.push(
+				_.defaults(header, { order: 0 })
+			);
+		},
+
+		registerFooter: function(header) {
+			this.footers.push(
+				_.defaults(header, { order: 0 })
```
Comment (danxuliu) [inline apps/files/js/filelist.js]: `header` => `footer`

**Review by danxuliu — CHANGES_REQUESTED** [review]

File: apps/files/js/filelist.js (hunk 2 above)
Comment (juliushaertl) [reply apps/files/js/filelist.js]: :see_no_evil: Thanks, fixed

**Review by ChristophWurst — APPROVED** [review]

**Review by danxuliu — APPROVED** [review]
Tested and works :+1: 

Note that the footer is a bit separated from the table due to [the summary height](https://github.com/nextcloud/server/blob/cbc38628302c618866d041f9b50958a25e16f838/apps/files/css/files.scss#L651-L652). I guess that it would be good to add that space only if there are no footers, although we can worry about that later too ;-)

**Comment by rullzer** [comment]
Fixup and we get it in @juliushaertl !

**Pull request merged**